File: rust/fluentcards/src/builders/image_set_builder.rs
use serde_json::Value;

use crate::enums::*;
use crate::models::Card;
use super::image_builder::ImageBuilder;

/// Builds an ImageSet Adaptive Card element.
pub struct ImageSetBuilder {
    data: Card,
}

impl ImageSetBuilder {
    pub(crate) fn new() -> Self {
        let mut data = Card::new();
        data.insert("type".into(), Value::String("ImageSet".into()));
        data.insert("images".into(), Value::Array(Vec::new()));
        ImageSetBuilder { data }
    }

    pub fn with_id(&mut self, id: &str) -> &mut Self {
        self.data.insert("id".into(), Value::String(id.into()));
        self
    }

    pub fn with_image_size(&mut self, size: ImageSize) -> &mut Self {
        self.data.insert("imageSize".into(), size.into());
        self
    }

    pub fn with_spacing(&mut self, spacing: Spacing) -> &mut Self {
        self.data.insert("spacing".into(), spacing.into());
        self
    }

    /// Adds an image configured by the provided closure.
    pub fn add_image(
        &mut self,
        configure: impl FnOnce(&mut ImageBuilder),
    ) -> &mut Self {
        let mut ib = ImageBuilder::new();
        configure(&mut ib);
        if let Some(Value::Array(images)) = self.data.get_mut("images") {
            images.push(Value::Object(ib.build()));
        }
        self
    }

    pub fn build(&self) -> Card {
        self.data.clone()
    }
}

File: swift/Sources/FluentCards/ImageSetBuilder.swift
/// ImageSetBuilder builds an ImageSet Adaptive Card element.
public final class ImageSetBuilder {
    var data: [String: Any] = ["type": "ImageSet", "images": [Any]()]

    public init() {}

    @discardableResult
    public func withID(_ id: String) -> ImageSetBuilder {
        data["id"] = id
        return self
    }

    @discardableResult
    public func withImageSize(_ size: ImageSize) -> ImageSetBuilder {
        data["imageSize"] = size.rawValue
        return self
    }

    @discardableResult
    public func withSpacing(_ spacing: Spacing) -> ImageSetBuilder {
        data["spacing"] = spacing.rawValue
        return self
    }

    /// Adds an image configured by the provided closure.
    @discardableResult
    public func addImage(_ configure: (ImageBuilder) -> Void) -> ImageSetBuilder {
        let ib = ImageBuilder()
        configure(ib)
        var images = data["images"] as! [Any]
        images.append(ib.build())
        data["images"] = images
        return self
    }

    public func build() -> Card {
        return data
    }
}
